### ISSUE TITLE
Feature: Support global taxonomy routes

### ISSUE THREAD

**Issue by jacksleight** [mitydigital/statamic-sitemapamic#1]
Currently collection specific taxonomy routes are supported, eg. `/blog/topics/design`, but global taxonomy routes aren't, eg. `/topics/design`. Be great if these could be added.

If you'd consider a PR I'd be happy to give it a go? I guess the `defaults` config would need to be expanded somehow to cover taxonomies as well as collections.

**Comment by martyf** [comment]
Hey @jacksleight that's a great idea. I've actually just done a push to the main repo (not published as a release yet) that does this.

The logic I've used here is:
- added a new `globals` config key, that has a `taxonomies` array for any you want included
- can optionally set the frequency and priority defaults, just like content
- the generator only includes the terms if the taxonomy has a matching view (i.e. will be a working URL, not a 404)
- gets cached with the rest of the sitemap, and cleared when the terms change

I thought separating to a new config key made sense just in case you had a collection called 'taxonomies', and because the behaviour is a bit different from entries.

Did you want to maybe install `dev-main` and give this a go to see what you think?

In the source, check out the config file for an example of the new config, and if you're curious, you can see the new behaviour in the `SitemapamicController`.

Oh and always happy for a PR in the future too 👍

**Comment by jacksleight** [comment]
Amazing! That's working great.

My only thought is could global terms also check the `meta_include_in_xml_sitemap` field? You might have some global terms that you still want to exclude.

Now you've added that global option I was thinking what else could go in there... The site I'm working on has some custom routes defined that are outside of Statamic, but I'd be great if I could get them into the sitemap. Only thing is while some of them are static and could be listed in the config, some have parameters so would need to be added dynamically somehow. Not sure if that's feasible, should probably be a separate FR.

**Comment by martyf** [comment]
No worries will get that added in the morning (9:30pm here atm)

Non-Statamic routes is a good idea too so will work on that as well. 

For the dynamic ones that need parameters, what sort of parameters? Did you have an example or two?

**Comment by jacksleight** [comment]
So I'm currently working on a site with an events collection, but outside of the usual collection route I have a bunch of registration pages set up with a custom Laravel route:

```php
Route::get('register/{event}', ...)->name('register')
```

I'd like to be able to add these pages to the sitemap, by looping through the events and adding each URL using the `route()` helper.

But these kind of custom routes could be literally anything, with all sorts of parameters, possibly not even involving Statamic data if it was being used as part of a larger Laravel app. The most flexible option would probably be some way to register a function with Sitemapamic that can be called when it's generating the sitemap. Maybe a bit like this?

```php
Sitemapamic::urls(function () {
    return $someCustomData->map(fn ($entry) => new SitemapUrl([
        'loc' => route('register', ['event' => $entry->slug]),
        'lastmod' => ...,
    ]));
})
```

The XML SItemap addon has a [similar feature](https://github.com/pecotamic/sitemap#dynamically-adding-entries-optional) for adding custom URLs.

Just an idea though! :-)

**Comment by martyf** [comment]
Both of these have been added - check out the `dev-main` for the latest updates.

I've also updated the README.md with more details about both Globals and Dynamic Routes.

For named routes (which is what I'd recommend too) you need to add your own Service Provider that is after the `RouteServiceProvider` in your app.php config file. There's a note about this in the readme. 

Also, just in case you used the `SitemapUrl` class at all, it has been renamed to be `SitemapamicUrl` and now accepts named attributes.

Let me know what you think!

**Comment by jacksleight** [comment]
This is awesome, thanks so much for adding these so quickly!

Just one last thing, if `Sitemapamic::addDynamicRoutes` takes an array then those URLs will be created on every request. If it took a closure instead they could just be created when Sitemapamic is generating a fresh sitemap?

**Comment by martyf** [comment]
Duh I didn't pick that up in my tests as I was just looking at the Sitemap.

Take a try again now, with a closure. The readme is updated too.

**Comment by jacksleight** [comment]
All working perfectly! 🎉

**Comment by martyf** [comment]
Hoorah! Thanks for your input! I’ll formalise and proof read the readme in the morning and tag the release.

**Comment by martyf** [comment]
All released in [2.1.0](./../releases/tag/2.1.0). I'll close this ticket off - thanks for your input and testing on this one!
